### DISCUSSION TITLE
How to implement messaging using IMessageBus

### DISCUSSION THREAD

**Discussion by natilivni** [dotnet/dotNext#43] [Q&A]
Hello,

I was wondering if I could get some guidance on how to implement messaging.  I have created a message which implements IMessage and have injected IMessageBus into my class.

I is clear to me that I now have to implement IInputChannel.  However it is not clear to me where and how I register such a handler in the code.

Further, I have noticed that when using the LeaderRouter property of the IMessageBus to send a message, I get a "no leader" error even though a leader has been elected.  Is there a place where I need to update the leader property manually?

**Comment by sakno** [comment]
If you want to use existing implementation then there is a lot of read-to-use `IMessage` implementations. In case of `DotNext.AspNetCore.Cluster` library, `LeaderRouter` just takes the address of the leader from `Leader` property and sends a message to it in a loop.

Message handler can be registered as a singleton service within DI of `IInputChannel` type.
